src/core/mempool.rs

```diff
@@ -1129,6 +1290,79 @@ impl MemPoolDB {
         Ok(updated)
     }
 
+    /// Counts and buckets the entire mempool DB into three buckets:
+    ///   * nonce too low
+    ///   * nonce ready to run
+    ///   * nonce in the future
+    pub fn bucket_count_candidates<C: ClarityConnection>(
+        &mut self,
+        clarity_tx: &mut C,
+        settings: MemPoolWalkSettings,
+    ) -> Result<HashMap<Ordering, u64>, db_error> {
+        debug!("Mempool walk to `bucket_candidates`");
+        let mut nonce_cache = NonceCache::new(settings.nonce_cache_size);
+        let sql = "
+            SELECT txid, origin_nonce, origin_address, sponsor_nonce, sponsor_address, fee_rate
+            FROM mempool
+            ";
+        let mut query_stmt = self
+            .db
+            .prepare(&sql)
+            .map_err(|err| Error::SqliteError(err))?;
+        let mut sql_iterator = query_stmt
+            .query(NO_PARAMS)
+            .map_err(|err| Error::SqliteError(err))?;
+
+        let mut result_map: HashMap<Ordering, u64> = HashMap::new();
+        loop {
+            let row = match sql_iterator.next().map_err(|err| Error::SqliteError(err))? {
+                Some(result) => result,
+                None => {
+                    break;
+                }
+            };
+
+            let tx_info = MemPoolTxInfoPartial::from_row(row)?;
+
+            // Check the nonces.
+            let expected_origin_nonce =
+                nonce_cache.get(&tx_info.origin_address, clarity_tx, self.conn());
+            let expected_sponsor_nonce =
+                nonce_cache.get(&tx_info.sponsor_address, clarity_tx, self.conn());
+            let nonce_order = order_nonces(
+                tx_info.origin_nonce,
+                expected_origin_nonce,
+                tx_info.sponsor_nonce,
+                expected_sponsor_nonce,
+            );
+
+            match result_map.get_mut(&nonce_order) {
+                Some(count) => {
+                    *count += 1;
+                }
+                None => {
+                    result_map.insert(nonce_order, 1);
+                }
+            }
+
+            // Bump nonces in the cache for the executed transaction
+            nonce_cache.update(
+                tx_info.origin_address,
+                expected_origin_nonce + 1,
+                self.conn(),
+            );
+            if tx_info.sponsor_address != tx_info.origin_address {
+                nonce_cache.update(
+                    tx_info.sponsor_address,
+                    expected_sponsor_nonce + 1,
+                    self.conn(),
+                );
+            }
```

Comment:
I don't think you should bump the nonces in here. We should just be looking at transactions that are ready now.